File: android/app/src/main/java/com/modi/connect/core/connectivity/channel/LegacyAudioChannelAdapter.kt
package com.modi.connect.core.connectivity.channel

import com.modi.connect.core.connectivity.AudioChannel
import com.modi.connect.core.connectivity.ChannelDataPlane
import com.modi.connect.core.connectivity.ChannelRuntimeState
import com.modi.connect.core.connectivity.SessionId
import com.modi.connect.core.connectivity.TransportSession
import java.util.concurrent.atomic.AtomicLong

/** Maps audio/primary to an approved legacy transport without inspecting or rewriting bytes. */
class LegacyAudioChannelAdapter(
    override val sessionId: SessionId,
    private val transport: TransportSession,
) : ChannelDataPlane {
    private val sequence = AtomicLong()
    override val descriptor = AudioChannel.descriptor
    override var state = ChannelRuntimeState.Closed
        private set
    override val nextSequence get() = sequence.get()
    override var onBytesReceived: ((ByteArray) -> Unit)? = null
    private var subscribed = false

    override suspend fun open() {
        if (!subscribed) {
            transport.onBytesReceived = { payload -> onBytesReceived?.invoke(payload) }
            subscribed = true
        }
        resetSequence()
        state = ChannelRuntimeState.Open
    }

    override fun reserveSequence(): Long {
        check(state == ChannelRuntimeState.Open) { "Channel is closed" }
        return sequence.getAndIncrement()
    }

    override fun resetSequence() = sequence.set(0)

    override suspend fun send(payload: ByteArray) {
        check(state == ChannelRuntimeState.Open) { "Channel is closed" }
        transport.send(payload)
    }

    override suspend fun close() {
        if (subscribed) { transport.onBytesReceived = null; subscribed = false }
        state = ChannelRuntimeState.Closed
    }
}

object AudioChannelComposition {
    fun isEnabled(readSetting: (String) -> String? = System::getProperty): Boolean =
        readSetting("modi.audio.channel") == "1"
}

File: windows/MoDi.Desktop/Connectivity/Channels/LegacyAudioChannelAdapter.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using MoDi.App.Contracts.Connectivity;

namespace MoDi.Desktop.Connectivity.Channels;

/// <summary>Maps audio/primary to an approved legacy transport without inspecting or rewriting bytes.</summary>
internal sealed class LegacyAudioChannelAdapter : IChannelDataPlane
{
    private readonly ITransportSession _transport;
    private long _nextSequence;
    private bool _subscribed;

    internal LegacyAudioChannelAdapter(SessionId sessionId, ITransportSession transport)
    {
        SessionId = sessionId;
        _transport = transport;
    }

    public SessionId SessionId { get; }
    public ChannelDescriptor Descriptor => AudioChannel.Descriptor;
    public ChannelRuntimeState State { get; private set; } = ChannelRuntimeState.Closed;
    public long NextSequence => Interlocked.Read(ref _nextSequence);
    public event Action<ReadOnlyMemory<byte>>? BytesReceived;

    public Task OpenAsync(CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        if (!_subscribed) { _transport.BytesReceived += OnBytesReceived; _subscribed = true; }
        ResetSequence();
        State = ChannelRuntimeState.Open;
        return Task.CompletedTask;
    }

    public long ReserveSequence()
    {
        if (State != ChannelRuntimeState.Open) throw new InvalidOperationException("Channel is closed.");
        return Interlocked.Increment(ref _nextSequence) - 1;
    }

    public void ResetSequence() => Interlocked.Exchange(ref _nextSequence, 0);

    public Task SendAsync(ReadOnlyMemory<byte> payload, CancellationToken cancellationToken)
    {
        if (State != ChannelRuntimeState.Open) throw new InvalidOperationException("Channel is closed.");
        return _transport.SendAsync(payload, cancellationToken);
    }

    public Task CloseAsync(CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        if (_subscribed) { _transport.BytesReceived -= OnBytesReceived; _subscribed = false; }
        State = ChannelRuntimeState.Closed;
        return Task.CompletedTask;
    }

    private void OnBytesReceived(ReadOnlyMemory<byte> payload) => BytesReceived?.Invoke(payload);
}
